sql-plugins/src/main/java/io/cdap/pipeline/sql/plugins/sources/BigQuerySQLSource.java

```diff
@@ -0,0 +1,162 @@
+import com.google.cloud.bigquery.Field;
+import com.google.cloud.bigquery.Schema;
+import com.google.cloud.bigquery.Table;
+import com.google.cloud.bigquery.TableId;
+import com.google.common.base.Strings;
+import io.cdap.cdap.api.annotation.Description;
+import io.cdap.cdap.api.annotation.Name;
+import io.cdap.cdap.api.annotation.Plugin;
+import io.cdap.cdap.api.plugin.PluginConfig;
+import io.cdap.pipeline.sql.api.template.SQLSource;
+import io.cdap.pipeline.sql.api.template.TableInfo;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.tools.RelBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+
+/**
+ * A BigQuery SQL source.
+ */
+@Plugin(type = SQLSource.PLUGIN_TYPE)
+@Name("BigQueryTable")
+@Description("A BigQuery source.")
+public class BigQuerySQLSource extends SQLSource {
+  private final BigQuerySQLSourceConfig config;
+  private static final Logger LOG = LoggerFactory.getLogger(BigQuerySQLSource.class);
+
+  public BigQuerySQLSource(BigQuerySQLSourceConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public RelNode getQuery(RelBuilder builder) {
+    return builder.build();
+  }
+
+  @Override
+  public TableInfo getSourceTable() throws IllegalArgumentException {
+    if (Strings.isNullOrEmpty(config.getProject())) {
+      throw new IllegalArgumentException("Destination project string must be provided.");
+    }
+    if (Strings.isNullOrEmpty(config.getDataset())) {
+      throw new IllegalArgumentException("Destination dataset string must be provided.");
+    }
+    if (Strings.isNullOrEmpty(config.getTable())) {
+      throw new IllegalArgumentException("Destination table string must be provided.");
+    }
+    TableInfo.Builder builder = TableInfo
+      .builder(String.format("%s.%s.%s", config.getProject(), config.getDataset(), config.getTable()));
+    // Get the schema from BigQuery API
+    GoogleCredentials credentials;
+    try {
+      File credentialsPath = new File(config.getServiceAccountPath());
+      FileInputStream serviceAccountStream = new FileInputStream(credentialsPath);
+      credentials = ServiceAccountCredentials.fromStream(serviceAccountStream);
+    } catch (IOException e) {
+      throw new IllegalArgumentException("Invalid credentials file");
+    }
+
+    // Instantiate a client.
+    BigQuery bigquery =
+      BigQueryOptions.newBuilder().setCredentials(credentials).build().getService();
+    TableId id = TableId.of(config.getDataset(), config.getTable());
+    Table bqTable = bigquery.getTable(id);
+    Schema bqSchema = bqTable.getDefinition().getSchema();
+    for (Field field: bqSchema.getFields()) {
+      String fieldName = field.getName();
+      String fieldTypeName = field.getType().toString().toUpperCase();
+      SqlTypeName calciteType = SqlTypeName.get(convertToCalciteType(fieldTypeName));
```

Comment:
does this work for array and struct fields?

Reply:
ARRAY is supported but STRUCT is not. I added a conversion definition in the below function for STRUCT -> STRUCTURED.